Lf 3106 details from drawn shape are not persisted when going back to draw step in flow

javiscript13 opened the pull request:
**Description**

This PR is for fixing the bug that whenever you were on the form for creating any area (Residence, Barn, etc), line (watercourse, buffer zone, etc) or marker (gate, water valve), and you clicked Back, and the confirmed again, the form was with its original values and not the ones that you have modified.

To fix this the upsertFormData form hookFormPersistSlice is now dispatched when the user clicks back on the form to return to the map.  

But the form has some calculated values (areas, perimeters for the polygons but also for some of the line types).  To be able to use the recalculates values if the user changed something (by using "Redraw" button, moving points or inserting new values for the line types) it was added to the store the isRedrawing slice. 

Some event listeners where added to the map in order to be able to detect when the user drags a point for a marker, polygon or line. And the also the click on the redraw button, or the onDirty for the preForm that is display for the line types.

Jira link: https://lite-farm.atlassian.net/browse/LF-3106

**Type of change**

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Passes test case
- [x] UI components visually reviewed on desktop view
- [ ] UI components visually reviewed on mobile view
- [ ] Other (please explain)

**Checklist:**

- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] The precommit and linting ran successfully
- [ ] I have added or updated language tags for text that's part of the UI
- [ ] I have added "MISSING" for all new language tags to languages I don't speak
- [ ] I have added [the GNU General Public License](https://lite-farm.atlassian.net/l/cp/BT0Dd7WW) to all new files

Duncan-Brain commented on packages/webapp/src/containers/Map/useDrawingManager.js:
```diff
@@ -75,6 +77,9 @@ export default function useDrawingManager() {
         ...getDrawingOptions(type).polygonOptions,
       });
       redrawnPolygon.setMap(map);
+      maps.event.addListener(redrawnPolygon.getPath(), 'set_at', () => {
```
I think this needs to be removed from the dom in the return statement of the useEffect.

Duncan-Brain commented on packages/webapp/src/containers/Map/useDrawingManager.js:
```diff
@@ -101,6 +106,9 @@ export default function useDrawingManager() {
         draggable: true,
       });
       redrawnMarker.setMap(map);
+      maps.event.addListener(redrawnMarker, 'dragend', () => {
```
same as above I think this needs to be removed from the dom in the return statement of the useEffect.

Duncan-Brain commented:
@javiscript13 

This looks like it was a tough one! It seems to be working well for me.

I have left 1 or 2 small inline comments but otherwise it looks good -- I am still learning the redux stuff but it looks like small changes.

javiscript13 commented:
> @javiscript13
> 
> This looks like it was a tough one! It seems to be working well for me.
> 
> I have left 1 or 2 small inline comments but otherwise it looks good -- I am still learning the redux stuff but it looks like small changes.

@Duncan-Brain it was a fun one.  Thank you for your comments, I have added the removeListeners for the two listeners that I introduced in the code.  I did have to add a condition to the cleanup function which is the thing I'm having doubts about. If I don't add that condition the listener will be removed right after it was added, so I'm thinking if that condition would be the "symmetrical" way to do it according to the docs (https://react.dev/reference/react/useEffect#my-cleanup-logic-runs-even-though-my-component-didnt-unmount).  

Please note that there is another listener on the useDrawingManager file that I'm not removing because the function that adds the listener is explicitly called from the Map where there are another bunch of listeners not being unmounted, I think that might be another task.

SayakaOno reviewed (commented):
`upsertFormData` solution is great!

I haven't fully understood why you needed to create `isRedrawing` state...
If I'm not mistaken, that's only used in `handleConfirm` and `handleLineConfirm`. What was the problem you saw with the old implementation?
(I wonder if we could always save the latest value whether it's the first time or not)

Duncan-Brain commented:
1. For the maps event listeners how do you view them? I cannot seem to find what I expect to see in the inspector. I am not very familiar with cleaning up after google maps so I am sorry if what I say isn't super helpful. Is there a special way to view google maps events? You mentioned:

> Please note that there is another listener on the useDrawingManager file that I'm not removing because the function that adds the listener is explicitly called from the Map where there are another bunch of listeners not being unmounted, I think that might be another task.

Are you talking about the addLineListeners function? If so I think this should be fixed as well. My understanding is that not removing them can cause unexpected behaviour or memory leaks. Here is how I would expect it to go but I am having trouble verifying:

```
if (isArea(type)) maps.event.removeListener(polygonDragListener);
if (isLine(type)) removeLineListeners() //to be added
if (isPoint(type)) maps.event.removeListener(markerDragListener);
```

But I like your check for the listener too and maybe yours is better. Its one of those things I am pretty sure needs to be done but I also not an expert since this can be tricky and just can't figure how to test it.

2. I am seeing the same thing as Sayaka on the adjusting line box.

javiscript13 commented on packages/webapp/src/containers/Map/useDrawingManager.js:
```diff
@@ -75,6 +78,9 @@ export default function useDrawingManager() {
         ...getDrawingOptions(type).polygonOptions,
       });
       redrawnPolygon.setMap(map);
+      polygonDragListener = maps.event.addListener(redrawnPolygon.getPath(), 'set_at', () => {
+        setPointChanged(true);
+      });
```
@SayakaOno Thank you for your testing, turns out that those "middle" points are not affected by the listeners I had added.  I added too the right listeners for those kind of points and it is now working.

SayakaOno replied:
Confirmed that it's working!
I would still love to learn about this:

> I haven't fully understood why you needed to create isRedrawing state...
If I'm not mistaken, that's only used in handleConfirm and handleLineConfirm. What was the problem you saw with the old implementation?
(I wonder if we could always save the latest value whether it's the first time or not)

SayakaOno replied:
Woops! I will read your reply now 🤣 Sorry!

javiscript13 commented:
> `upsertFormData` solution is great!
> 
> I haven't fully understood why you needed to create `isRedrawing` state... If I'm not mistaken, that's only used in `handleConfirm` and `handleLineConfirm`. What was the problem you saw with the old implementation? (I wonder if we could always save the latest value whether if it's the first time or not)

Thank you @SayakaOno!  . 

about isRedrawing:

Yes, it is only being checked on the "confirm" button for polygons and points (both in `handleConfirm`) and lines (`handleLineConfirm`).

So the problem that I'm solving by using it is the following:  

When the users draws the shape and then press confirm button and goes to the form, every point data and calculated values such as area and perimeter are stored on the redux form.  So, if the user click back and redraws something (by clicking "redraw" button or just changing the points as in your video) the new values are inserted on the form which is great.  But, If the users manually changes the area or perimeter value on the form and then presses < back, does not changes anything and click "confirm" again the area or perimeter value that the user has manually changed will have the calculated value again.  So, I'm using isRedrawing to know when those values should be overwritten (only if the user changes something on the drawing, if user has not changed the drawing the value that has previously filled will be available).   I don't think that we can compare the calculated value with what is on the redux stored value, because the value will be differente when the user has changed something in the drawing, but it will be different when the user has changed something on the form too.

Please let me know if you find this explanation useful, and if there is a better way to handle it.

javiscript13 commented:
> 1. For the maps event listeners how do you view them? I cannot seem to find what I expect to see in the inspector. I am not very familiar with cleaning up after google maps so I am sorry if what I say isn't super helpful. Is there a special way to view google maps events? You mentioned:
> 
> > Please note that there is another listener on the useDrawingManager file that I'm not removing because the function that adds the listener is explicitly called from the Map where there are another bunch of listeners not being unmounted, I think that might be another task.
> 
> Are you talking about the addLineListeners function? If so I think this should be fixed as well. My understanding is that not removing them can cause unexpected behaviour or memory leaks. Here is how I would expect it to go but I am having trouble verifying:
> 
> ```
> if (isArea(type)) maps.event.removeListener(polygonDragListener);
> if (isLine(type)) removeLineListeners() //to be added
> if (isPoint(type)) maps.event.removeListener(markerDragListener);
> ```
> 
> But I like your check for the listener too and maybe yours is better. Its one of those things I am pretty sure needs to be done but I also not an expert since this can be tricky and just can't figure how to test it.
> 
> 2. I am seeing the same thing as Sayaka on the adjusting line box.

Thank you @Duncan-Brain !

I also have trouble finding google map api events on the developer tools inspector. I know they are there for the code:

this is containers/Map/index.js:

![image](https://user-images.githubusercontent.com/4588278/235798577-e5031415-b34a-4a1e-ba92-6a08f7517337.png)

Those and more listeners are added and there isn't a place where they are removed. (Which I understand is necessary, but suggesting to be a ticket on its own)

About the line listener. In the `useDrawingManager.js` file the `addLineListeners` function is being called in two places: 

1. In the same place where I'm creating the other listeners, and 
2. the `finishDrawing` function, and that function is being called form Map/index.js:

`useDrawingManager.js`:
![image](https://user-images.githubusercontent.com/4588278/235799176-358bc50a-3457-4a5a-a285-35412a63247a.png)

called from (Map/index.js):
![image](https://user-images.githubusercontent.com/4588278/235799239-aea0f5a8-ac62-4855-b14c-d851efa2ee46.png)

So I'm not sure about the right way to proceed with it since I think it might affect other functionality already working (but that might need some rework).

I have fixed the issue mentioned by Sayaka, you can read more details about it on the comment where she pointed it out.

SayakaOno commented:
@javiscript13 I see!!!!!!!!! So that was not just the first time VS the second+ time comparison! Hats off to you!!

Duncan-Brain commented:
@javiscript13 I am good with a new Jira ticket for all cleanup then. Since we can't verify we are removing them right now we could add it to that ticket instead.

javiscript13 commented:
> @javiscript13 I am good with a new Jira ticket for all cleanup then. Since we can't verify we are removing them right now we could add it to that ticket instead.

thank you @Duncan-Brain , I have created the ticket:  https://lite-farm.atlassian.net/browse/LF-3241

Duncan-Brain reviewed (approved):
@javiscript13 

Thank you so much Javier looks great!
